### PR TITLE
feat(mirror): add Jihulab as optional git repo

### PR DESCRIPTION
**Pull request by leeebo**
This PR aims to add another git mirror for users to choose

[Jihulab](https://jihulab.com/) is a Chinese localization of Gitlab public repository,

Espressif maintained an official mirror there https://jihulab.com/esp-mirror/espressif/esp-idf,

Which has the same submodule relative paths as the Github repo.

### CODE REVIEW

File: src/InnoSetup/Summary.iss [hunk 1]
```diff
@@ -114,6 +114,8 @@ begin
   end;
   if (WizardIsComponentSelected('{#COMPONENT_OPTIMIZATION_GITEE_MIRROR}')) then begin
     Result := Result + 'Mirror (Gitee.com); ';
+  end else if (WizardIsComponentSelected('{#COMPONENT_OPTIMIZATION_JIHULAB_MIRROR}')) then begin
```
Comment (jakub-kocka) [inline src/InnoSetup/Summary.iss:117]: Is there a special need for the usage of the `else if` construction?

File: src/InnoSetup/Languages/IdfToolsSetup_sk-SK.isl [hunk 2]
```diff
@@ -156,6 +156,7 @@ SystemVersionTooLow=Príliš stará verzia operačného systému. Použite podpo
 WindowsVersion=Verzia Windows
 SystemCheckAlternativeMirror=Test alternatívneho zrkadla
 ComponentOptimizationGiteeMirror=Git zrkadlo gitee.com/EspressifSystems/esp-idf
+ComponentOptimizationJihulabMirror=Git zrkadlo jihulab.com/esp-mirror/espressif/esp-idf/
```
Comment (jakub-kocka) [inline src/InnoSetup/Languages/IdfToolsSetup_sk-SK.isl:159]: We should change this as well to mirror (zrkadlo -> mirror)

File: src/InnoSetup/Languages/IdfToolsSetup_cs-CZ.isl [hunk 3]
```diff
@@ -156,6 +156,7 @@ SystemVersionTooLow=Příliš stará verze operačního systému. Použijte podp
 WindowsVersion=Verze Windows
 SystemCheckAlternativeMirror=Test alternativního zrcadla
 ComponentOptimizationGiteeMirror=Git zrcadlo gitee.com/EspressifSystems/esp-idf
+ComponentOptimizationJihulabMirror=Git zrcadlo jihulab.com/esp-mirror/espressif/esp-idf/
```
Comment (jakub-kocka) [inline src/InnoSetup/Languages/IdfToolsSetup_cs-CZ.isl:159]: We should change this to mirror (zrcadlo -> mirror) in all occurrences

**Review by jakub-kocka — COMMENTED** [review]
Thanks @leeebo for the PR proposal, I have left a few comments. Otherwise, it looks good to me.

But I have noticed that the `Jihulab option` is checked out by default, I think we should update this for the default version of the Installer to be not checked.

**Comment by leeebo** [comment]
> One more thing, maybe we should have `Jihulab` and `Gitee` dependent on each other, and when one option is set the other one will not be possible to set.

Yes, make sense. I want to do that, but I don't know the syntax, that's why I write `else if` [here](https://github.com/espressif/idf-installer/pull/249#discussion_r1647385036)

Will fix it

File: src/InnoSetup/IdfToolsSetup.iss [hunk 4]
```diff
@@ -326,6 +327,7 @@ Name: "{#COMPONENT_TARGET_ESP32_S3}"; Description: {cm:ComponentTargetEsp32s3};
 #if OFFLINE == 'no'
 Name: "{#COMPONENT_OPTIMIZATION}"; Description: {cm:ComponentOptimization}; Types: custom;
 Name: "{#COMPONENT_OPTIMIZATION_ESPRESSIF_DOWNLOAD}"; Description: {cm:ComponentOptimizationEspressifDownload}; Types: full custom; Flags: checkablealone
+Name: "{#COMPONENT_OPTIMIZATION_JIHULAB_MIRROR}"; Description: {cm:ComponentOptimizationJihulabMirror}; Types: custom; Flags: checkablealone
```
Comment (leeebo) [inline src/InnoSetup/IdfToolsSetup.iss:330]: `But I have noticed that the Jihulab option is checked out by default, I think we should update this for the default version of the Installer to be not checked.`

@jakub-kocka I think it is already not checked. Please correct me if I was wrong

Comment (jakub-kocka) [reply src/InnoSetup/IdfToolsSetup.iss:330]: I have had the same impression from the code but when tested on the built installer it was checked.

I will take a look at this more.

Comment (leeebo) [reply src/InnoSetup/IdfToolsSetup.iss:330]: Ok, I got it, you may currently in the [leeebo/main](https://github.com/leeebo/idf-installer/tree/main) branch which changes the default value (for my test), but the PR is from [feat/add_jihulab_git_mirror](https://github.com/leeebo/idf-installer/tree/feat/add_jihulab_git_mirror)

Comment (jakub-kocka) [reply src/InnoSetup/IdfToolsSetup.iss:330]: You are right the `Types: custom;` is correct and then it is not checked.

**Comment by jakub-kocka** [comment]
> > One more thing, maybe we should have `Jihulab` and `Gitee` dependent on each other, and when one option is set the other one will not be possible to set.
> 
> Yes, make sense. I want to do that, but I don't know the syntax, that's why I write `else if` [here](https://github.com/espressif/idf-installer/pull/249#discussion_r1647385036)
> 
> Will fix it

This should be only for summary text so it is OK to put there just `if`.
TBH, I am unsure about the syntax as well, but if you need some help, I can dig a bit more into how to implement this.

**Comment by leeebo** [comment]
@jakub-kocka Now users can exclusively check or uncheck the `Git Mirror` like below: 

English Language:
image

Chinese Language:
07fb808481b20b619d8f684a446a661

I have tested on my  Windows11 PC, it works as expected, the log as bellow: 
[Setup Log 2024-06-27 #003.txt](https://github.com/user-attachments/files/16007103/Setup.Log.2024-06-27.003.txt)

However, it should be noted that since the submodule of the submodule may specify the URL from github directly (like below, release/v5.3, cmock's submodules are from github), even if the user chooses Git mirror, these submodules still download from Github. As suggested from @wujiangang, we may need to run command like `git config - global url.https://jihulab.com/esp-mirror.insteadOf https://github.com ` to fix it.

image

**Comment by jakub-kocka** [comment]
@leeebo Nice work! Thank you for the improvements.

I agree with changing the git URL. If it may wait a bit longer we can fix this in the scope of the task with `git clone` fix (internal tracker IDF-10228 - it is planned for Q3) so this PR can be eventually merged.

**Comment by leeebo** [comment]
@jakub-kocka Comment has been added in the jira, let's wait for your fix

**Comment by jakub-kocka** [comment]
> @jakub-kocka Comment has been added in the jira, let's wait for your fix

@leeebo, thank you.

So this PR is ready to merge, right?

**Comment by dobairoland** [comment]
@leeebo Do you think this mirror would be useful to mention at https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/linux-macos-setup.html#step-2-get-esp-idf? We have later on that page Github mirror for downloadable tools but no alternative mentioned to cloning from Github.

**Comment by jakub-kocka** [comment]
Thank you @dobairoland for the review.
I am merging this and will release the latest IDF version offline installer and online installer.

**Comment by leeebo** [comment]
@dobairoland Thanks for your suggestions, I think finally we should mention the mirrors at https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/linux-macos-setup.html#step-2-get-esp-idf, we currently need an additional step to change the submodule's submodule URL from github to jihulab.

How about adding a new script like `tools/set-submodules-to-jihulab.sh` ?  in esp-idf

**Comment by dobairoland** [comment]
> we currently need an additional step to change the submodule's submodule URL from github to jihulab.

No, I don't think that is necessary. We use relative paths here: https://github.com/espressif/esp-idf/blob/master/.gitmodules. That means that the submodules URLs will point to the same page domain the root repository was cloned from. The same way it works for our Github & Gitlab. I don't see why it wouldn't work for other mirrors as well.

**Comment by leeebo** [comment]
@dobairoland For the ESP-IDF's submodules it's OK, but for submodules's submodule, eg. CMock has submodule `vendor/unity` and `vendor/unity`, which using specified URL instead of relative paths. https://github.com/ThrowTheSwitch/CMock/blob/eeecc49ce8af123cf8ad40efdb9673e37b56230f/.gitmodules

**Comment by dobairoland** [comment]
Ok, I see. I wouldn't add a script to ESP-IDF which is not even needed directly by it. But I'm sure we can have such helper script in another repository.

Depending on the number of such submodules we can consider to submit PRs into them.

**Pull request merged**